docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,210 @@
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters (in any namespace).
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce"
+ clusterSelector:
+   matchLabels:
+     postcreatelabelcni: calico
+ resources:
+   - name: calico-addon
+     kind: Secret
+   - name: network-policy-addon
+     kind: ConfigMap
+status:
+ LastUpdated:  "2020-05-05T08:24:17Z"
+ StatusMessages: {
+                    {Message: "Added calico-addon resource", Type: "Update"},
+                    {Message: "Failed applying to x cluster", Type: "Error"},
+                }
+```
+
+Initially, the only supported mode will be `ApplyOnce` and it will be the default mode if no mode is provided. In the future, we may consider adding a `Sync` mode that reapplies the resources on resource hash change and/or periodically.
+If ClusterResourceSet resources will be managed by an operator after they are applied by ClusterResourceSet controller, "ApplyOnce" mode must be used so that reconciliation on those resources can be delegated to the operator.
+
+
+Each item in the resources specifies a kind (must be either ConfigMap or Secret) and a name. Each referenced ConfigMap/Secret  contains yaml/json content as value. The key to that content is “value”.
+
+*Sample Secret Format*
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name:calico-addon
+type: Opaque
+stringData:
+  value: |-
+    kind: ConfigMap
+    apiVersion: v1
+    metadata:
+     name: calico-conf
+```
+
+The resources in ClusterResourceSet will be applied to matching clusters. 
+A configmap will be created in the management cluster to keep track of which resources are applied by ClusterResourceSet resources. There will be one configmap per workload cluster.
+
+Example:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: capi-crs-my-cluster
+  namespace: capi-crs-my-cluster-namespace
+data:
+  <ClusterResourceSet-name1>:   
+    <secret-name1>:
+      hash: <>
+      status: success
+      error: ""
+      lastAppliedTime: "2020-04-05T08:24:17Z"
+    <configmap-name1>:
+      hash: <>
+      status: failed
+      error: "some error"
+      lastAppliedTime: "2020-05-05T08:24:17Z"
+  <ClusterResourceSet-name2>:  
+    <secret-name2>:
+      hash: <>
+      status: success
+      error: ""
+      lastAppliedTime: "2020-04-05T08:24:17Z"
+  Status: InProgress/Completed
```

Comment:
Is this duplicated from the ClusterResourceSet's Status?

Reply:
No, it is not. ClusterResourceSet do not have a status field for this. Keeping status per cluster gives more insight IMO.

Reply:
Having a CRD for keeping track of status. So this is no longer a concern.